cooperative_lanechange/src/cooperative_lanechange.cpp

```diff
@@ -0,0 +1,867 @@
+        current_gap = remaining_downtrack;
+
+
+        return current_gap;
+
+    }
+
+    void CooperativeLaneChangePlugin::pose_cb(const geometry_msgs::PoseStampedConstPtr& msg)
+    {
+        pose_msg_ = geometry_msgs::PoseStamped(*msg.get());
+    }
+    void CooperativeLaneChangePlugin::twist_cd(const geometry_msgs::TwistStampedConstPtr& msg)
+    {
+        current_speed_ = msg->twist.linear.x;
+    }
+    
+    void CooperativeLaneChangePlugin::bsm_cb(const cav_msgs::BSMConstPtr& msg)
+    {
+        bsm_core_ = msg->core_data;
+        
+    }
+
+    void CooperativeLaneChangePlugin::run()
+    {
+    	initialize();
+        ros::CARMANodeHandle::setSpinRate(10.0);
+        ros::CARMANodeHandle::spin();
+
+    }
+
+    
+
+    bool CooperativeLaneChangePlugin::plan_trajectory_cb(cav_srvs::PlanTrajectoryRequest &req, cav_srvs::PlanTrajectoryResponse &resp){
+        
+        //Check if Lane Change is needed
+        std::vector<cav_msgs::Maneuver> maneuver_plan;
+        for(const auto maneuver : req.maneuver_plan.maneuvers)
+        {
+            if(maneuver.type == cav_msgs::Maneuver::LANE_CHANGE)
+            {
+                maneuver_plan.push_back(maneuver);
+            }
+        }
+
+        //Current only checking for first lane change maneuver message
+        int target_lanelet_id = stoi(maneuver_plan[0].lane_change_maneuver.ending_lane_id);
+        double target_downtrack = maneuver_plan[0].lane_change_maneuver.end_dist;
+        //get subject vehicle info
+        lanelet::BasicPoint2d veh_pos(req.vehicle_state.X_pos_global, req.vehicle_state.Y_pos_global);
+        double current_downtrack = wm_->routeTrackPos(veh_pos).downtrack;
+        auto current_lanelets = lanelet::geometry::findNearest(wm_->getMap()->laneletLayer, veh_pos, 10);       
+        int current_lanelet_id = current_lanelets[0].second.id();
+        if(current_lanelet_id == target_lanelet_id && current_downtrack >= target_downtrack - destination_range_){
+            cav_msgs::LaneChangeStatus lc_status_msg;
+            lc_status_msg.status = cav_msgs::LaneChangeStatus::PLANNING_SUCCESS;
+            //No description as per UI documentation
+            lanechange_status_pub_.publish(lc_status_msg);
+        }
+
+        std::vector<cav_msgs::RoadwayObstacle> rwol = wm_->getRoadwayObjects();
+        //Assuming only one object in list 
+        int veh2_lanelet_id = rwol[0].lanelet_id;
+        double veh2_downtrack = rwol[0].down_track;  //Returns downtrack
+        double veh2_speed = rwol[0].object.velocity.twist.linear.x;
+
+        //get current_gap
+        double current_gap = find_current_gap(veh2_lanelet_id,veh2_downtrack);
+        //get desired gap - desired time gap (default 3s)* relative velocity
+        double relative_velocity = current_speed_ - veh2_speed;
+        double desired_gap = desired_time_gap_ * relative_velocity;      
+         
+        if(current_gap < desired_gap){
+            return true;
```

Comment:
If we have enough gap we exit the trajectory plan callback without any other planning being done?  Is the expectation then that this plugin should *only* be invoked if there is a definite conflict between the host vehicle and a known other vehicle in the target lane?

Reply:
The planning is being done regardless of there being enough gap. It's only when the gap is large enough that the plan is passed to the response. 
There doesn't have to be a definite conflict, but this plugin is different from unobstructed lanechange it doesn't always return a trajectory plan when its called. It will look for that conflict first.
I changed the default for "is_lanechange_accepted" to true, so if there's no other vehicle found, it should work like unobstructed lc